lib/Backends/Interpreter/InterpreterFunction.h

```diff
@@ -60,166 +54,164 @@ class InterpreterFunction final : public CompiledFunction {
 
   /// Allocate an unowned tensor to back the value \p v. The source tensor of
   /// the unowned tensor is provided by \p src.
   /// \returns a tensor for \p v.
-  Tensor *getOrCreateUnownedTensor(const Value *v, const Value *src,
+  Tensor *getOrCreateUnownedTensor(const Value *v, Context *ctx,
+                                   const Value *src,
                                    llvm::ArrayRef<size_t> offsets);
 
   /// If a tensor is allocated for \p v then delete it.
-  void deleteTensor(const Value *v);
+  void deleteTensor(const Value *v, Context *ctx);
 
   /// \returns a typed handle to the tensor that is stored at \p v.
   template <class ElemTy = float>
-  Handle<ElemTy> getWeightHandle(Value *v) const {
-    return getTensor(v)->getHandle<ElemTy>();
+  Handle<ElemTy> getWeightHandle(Value *v, Context *ctx) const {
+    return getTensor(v, ctx)->getHandle<ElemTy>();
   }
 
+  /// Loads constants from the runtime bundle into the context without
+  /// allocating new buffers.
+  void loadConstantsIntoContext(Context *ctx) const;
+
   /// @name Interpreter methods. This is a list of method declerations that are
   /// used by the interpreter to dispatch different instructions.
   ///@{
 
 #define DEF_VALUE(CLASS, NAME)
-#define DEF_INSTR(CLASS, NAME) void fwd##CLASS(const CLASS *I);
+#define DEF_INSTR(CLASS, NAME) void fwd##CLASS(const CLASS *I, Context *ctx);
 #define DEF_BACKEND_SPECIFIC_INSTR(CLASS, NAME)
 #include "glow/AutoGenInstr.def"
 
   template <typename ElemTy, typename AccumulatorTy>
-  void fwdConvolutionInstQuantizedImpl(Value *inV, Value *outV, Value *filterV,
-                                       Value *biasV,
+  void fwdConvolutionInstQuantizedImpl(Context *ctx, Value *inV, Value *outV,
+                                       Value *filterV, Value *biasV,
                                        llvm::ArrayRef<unsigned_t> kernelSizes,
                                        llvm::ArrayRef<unsigned_t> strides,
                                        llvm::ArrayRef<unsigned_t> pads,
                                        size_t group);
 
   template <typename ElemTy = float>
-  void fwdConvolutionInstFloatImpl(Value *inV, Value *outV, Value *filterV,
-                                   Value *biasV,
+  void fwdConvolutionInstFloatImpl(Context *ctx, Value *inV, Value *outV,
+                                   Value *filterV, Value *biasV,
                                    llvm::ArrayRef<unsigned_t> kernelSizes,
                                    llvm::ArrayRef<unsigned_t> strides,
                                    llvm::ArrayRef<unsigned_t> pads,
                                    size_t group);
 
-  void fwdAvgPoolInstI8Impl(const AvgPoolInst *I);
-  template <typename ElemTy> void fwdAvgPoolInstFloatImpl(const AvgPoolInst *I);
-  template <typename ElemTy> void fwdSoftMaxInstImpl(const SoftMaxInst *I);
+  void fwdAvgPoolInstI8Impl(const AvgPoolInst *I, Context *ctx);
```

Comment:
Definitely agree that would be less invasive, but I think also it will be less clear about what's actually happening. I'll put up another commit with an implementation that looks like that and we can see.